src/main/java/bio/terra/service/resourcemanagement/google/GoogleResourceService.java

```diff
@@ -188,26 +171,37 @@ public GoogleBucketResource getOrCreateBucket(GoogleBucketRequest bucketRequest,
             } else {
                 // bucket EXISTS. if this is a recovery flight or testing environment, use it. otherwise throw exception
                 // testing environments should set datarepo.gcs.allowReuseExistingBuckets=true in application.properties
-                if (metadataCreationFailed || allowReuseExistingBuckets) {
-                    logger.debug(String.format("bucket already exists, using anyway: %s", bucketName));
+                if (allowReuseExistingBuckets) {
```

Comment:
Couldn't a recovery flight in a non-testing environment (i.e. allowReuseExistingBuckets=false) hit the else case here, throwing an exception and deleting the bucket metadata, instead of just unlocking and returning the already-created bucket? I think that you need the metadataCreationFailed variable to know whether that's the case (i.e. the bucket metadata row is already locked by this flight, the bucket cloud resource already exists, so we just need to unlock and carry on). For this reason, my preference is to separate the createAndLock and get into separate dao methods, rather than having a single getOrCreateAndLockBucket method.

Reply:
Maybe we have too many cases in one place. The logic in this method was hard to understand when I reviewed your version. In trying to make it simpler - at least the way my brain works - maybe I missed a case. If it is that complex, maybe we need to further separate the cases so that the code is more obviously correct.

Let me write down all of the cases in a separate doc and make sure I have them all down. Then I'll transfer that into code comments and take another try at this.

Reply:
https://docs.google.com/document/d/1RLFHyNwXjQKjTlgD_e3G3cCNvZzHq1c_mqv1U2LWOXE/edit#